### PR TITLE
Fix SearchBar layout for big leadingView and big text mode

### PR DESCRIPTION
**Pull request by nemOoO-wang**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

1. Add limitation to leadingView width inside `SearchBar`, to fix layout break and interaction disability when leadingView too big or in accessibility big text mode
2. Add `lineBreakMode` to `BadgeView` and fix several indents
3. Add avatar image to SearchBar Demo

### Verification

> Some of test cases are in accessibility big text mode

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| a-long-name-old | a-long-name-new2 |
| a-long-name-with-avatar-old | a-long-name-with-avatar-new2 |
| long-name-old | long-name-new2 |
| long-name-with-avatar-old | long-name-with-avatar-new2 |

### Pull request checklist

This PR has considered:
- [ ] Light and Dark appearances
- [ ] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/1053)

### CODE REVIEW

File: ios/FluentUI/Navigation/SearchBar.swift [hunk 1]
```diff
@@ -454,7 +455,9 @@ open class SearchBar: UIView {
         }
 
         searchTextFieldBackgroundView.addSubview(leadingView)
-        let leadingViewSize = leadingView.sizeThatFits(searchTextFieldBackgroundView.frame.size)
+        let leadingViewRenderSize = CGSize(width: searchTextFieldBackgroundView.frame.size.width - (Constants.searchIconImageViewDimension + Constants.searchIconInset + Constants.searchTextFieldLeadingInset + Constants.searchTextFieldInteractionMinWidth + Constants.clearButtonLeadingInset + Constants.clearButtonWidth + Constants.clearButtonTrailingInset),
```
Comment (laminesm) [inline ios/FluentUI/Navigation/SearchBar.swift:458]: Would it make sense to compute all the constant values in a variable for readability?

File: ios/FluentUI.Demo/FluentUI.Demo.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -258,6 +258,7 @@
 				5340828A26CFF298007716E1 /* Recovered References */,
 			);
 			sourceTree = "<group>";
+			usesTabs = 0;
```
Comment (sophialee0416) [inline ios/FluentUI.Demo/FluentUI.Demo.xcodeproj/project.pbxproj:261]: Why is this change relevant to this PR?

File: ios/FluentUI.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -786,6 +786,7 @@
 				923DF2E62712B6AB00637646 /* Frameworks */,
 			);
 			sourceTree = "<group>";
+			usesTabs = 0;
```
Comment (sophialee0416) [inline ios/FluentUI.xcodeproj/project.pbxproj:789]: (here as well)

File: ios/FluentUI.Demo/FluentUI.Demo.xcodeproj/project.pbxproj (hunk 2 above)
Comment (nemOoO-wang) [reply ios/FluentUI.Demo/FluentUI.Demo.xcodeproj/project.pbxproj]: sorry, maybe it's auto updated by my local Apex config, reverted

File: ios/FluentUI/Navigation/SearchBar.swift [hunk 4]
```diff
@@ -144,6 +145,13 @@ open class SearchBar: UIView {
         static let navigationBarTransitionHidingDelay: TimeInterval = 0.5
 
         static let defaultStyle: Style = .lightContent
+
```
Comment (sophialee0416) [inline ios/FluentUI/Navigation/SearchBar.swift:148]: https://github.com/microsoft/fluentui-apple/runs/7220397515?check_suite_focus=true#step:5:147
Running pod lib lint should also catch these errors if your linter isn't

Comment (nemOoO-wang) [reply ios/FluentUI/Navigation/SearchBar.swift]: cool, validated using`./scripts/validation.sh`

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/SearchBarDemoController.swift [hunk 5]
```diff
@@ -13,11 +13,18 @@ class SearchBarDemoController: DemoController, SearchBarDelegate {
     }()
 
     private lazy var badgeView: UIView = {
-        let dataSource = BadgeViewDataSource(text: "Kat Larsson")
+        let imageView = UIImageView(image: UIImage(named: "avatar_kat_larsson"))
```
Comment (harrieshin) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/SearchBarDemoController.swift:16]: we should have both cases in the test app which has avatar image in the badgeview and not have avatar in the bage

File: ios/FluentUI/Navigation/SearchBar.swift [hunk 6]
```diff
@@ -135,6 +135,7 @@ open class SearchBar: UIView {
         static let searchIconInset: CGFloat = 10.0
         static let searchTextFieldLeadingInset: CGFloat = 10.0
         static let searchTextFieldVerticalInset: CGFloat = 2
+        static let searchTextFieldInteractionMinWidth: CGFloat = 50
```
Comment (harrieshin) [inline ios/FluentUI/Navigation/SearchBar.swift:138]: where does this come from?

**Comment by harrieshin** [comment]
what is native iOS textfield do in this case of larger text? like in mail.app or imessage.app when you try to put contact? should textfield text size increase at all? or textfield itself need to be multiline scrollable?

File: ios/FluentUI/Badge Field/BadgeView.swift [hunk 7]
```diff
@@ -385,12 +394,13 @@ open class BadgeView: UIView {
...
+            let customViewOrigin = CGPoint(x: customViewPadding.left, y: (frame.height - customViewSize.height) / 2)
             dataSource?.customView?.frame = CGRect(origin: customViewOrigin, size: customViewSize)
-			let labelOrigin = CGPoint(x: customViewPadding.left + customViewPadding.right + customViewSize.width, y: (frame.height - labelSize.height) / 2)
-            label.frame = CGRect(origin: labelOrigin, size: labelSize)
+            let labelOrigin = CGPoint(x: customViewPadding.left + customViewPadding.right + customViewSize.width, y: (frame.height - labelSize.height) / 2)
+            let labelSizeThatFits = CGSize(width: frame.size.width - labelOrigin.x, height: labelSize.height)
+            label.frame = CGRect(origin: labelOrigin, size: labelSizeThatFits)
         } else {
-            label.frame = bounds.insetBy(dx: -size.horizontalPadding, dy: -size.verticalPadding)
+            label.frame = bounds.insetBy(dx: size.horizontalPadding, dy: size.verticalPadding)
```
[the remark this answers is not in the document]
Comment (nemOoO-wang) [reply ios/FluentUI/Badge Field/BadgeView.swift:403]: it's required on big text mode, otherwise the layout will overlap over border

| before | after  |
| ------------- | ------------- |
| a-default-inset | a-updated-inset |
| default-inset | updated-inset |

File: ios/FluentUI/Navigation/SearchBar.swift (hunk 6 above)
Comment (nemOoO-wang) [reply ios/FluentUI/Navigation/SearchBar.swift:138]: actually it comes from my tests on DemoVC, we don't have spec from designer of this minimum width
if we need their confirm, who can I find?

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/SearchBarDemoController.swift [hunk 8]
```diff
@@ -30,6 +31,18 @@ class SearchBarDemoController: DemoController, SearchBarDelegate {
...
+    }
+
+    func buildBadgeView(text: String, customView: UIView? = nil) -> UIView {
+        let dataSource = BadgeViewDataSource(text: text, customView: customView)
+        let badge = BadgeView(dataSource: dataSource)
+        badge.lineBreakMode = .byTruncatingTail
+        badge.disabledBackgroundColor = Colors.Palette.blueMagenta20.color
+        badge.disabledLabelTextColor = .white
+        badge.isActive = false
+        badge.maxFontSize = 40
```
Comment (laminesm) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/SearchBarDemoController.swift:44]: Can we store these values in a Constants struct? Would make it easier to track and know where they're used.

**Comment by nemOoO-wang** [comment]
> what is native iOS textfield do in this case of larger text? like in mail.app or imessage.app when you try to put contact? should textfield text size increase at all? or textfield itself need to be multiline scrollable?

@harrieshin native UITextField cannot be multiline (UITextView is able), the native behavior is the same on iMessage and Settings app in big text mode

File: ios/FluentUI/Navigation/SearchBar.swift (hunk 6 above)
Comment (nemOoO-wang) [reply ios/FluentUI/Navigation/SearchBar.swift:138]: discussed with our team, we think this width is a protection when the text is too big or too long and let users to be able to interact. it not represented in design and doesn't have to be a precise value, and the `50` here is good enough for user to tap on textfield to edit.

**Comment by nemOoO-wang** [comment]
@sophialee0416 @harrieshin @mischreiber  gently ping

**Comment by laminesm** [comment]
> @sophialee0416 @harrieshin @mischreiber gently ping

Apologies for the delay. We're getting a designer to look at the spacing change you made. We'll get back to you in a short while.

**Comment by nemOoO-wang** [comment]
Hi @laminesm ，do we have any updates？
is there anything I could help? 😃 

cc @sophialee0416 @harrieshin @mischreiber

**Comment by redongreen** [comment]
@laminesm The suggestion is reasonable and does address the spacing issues we have with larger text. The screen below shows the min/max width added to fix UX by allowing users with larger text to type in the search field.

#### Why do we have this issue in the first place? 
This issue could have been prevented with detailed behavior specs. We do not have behavior specifications for searchbar and badgeview components; I expect we will have similar issues with other components.

For future consideration, we will need to provide (when applicable) specifications with min and max-width accounting for larger screen sizes and accessibility settings. 

@nemOoO-wang Thank you for the suggestions! We appreciate your contribution to Fluent!

**Review by laminesm — APPROVED** [review]

**Comment by laminesm** [comment]
> Hi @laminesm ，do we have any updates？ is there anything I could help? 😃
> 
> cc @sophialee0416 @harrieshin @mischreiber

Thank you for your patience and contribution!
